File: hackerrank/30 days of code/day 29, bitwise and/bitwise_and.go
// https://www.hackerrank.com/challenges/30-bitwise-and/problem?isFullScreen=true

package main

import "fmt"

func main() {
    var testCases int
    fmt.Scan(&testCases)
    for range testCases {
        var n, k int
        fmt.Scan(&n, &k)
        fmt.Println(bitwiseAnd(n, k))
    }
}

func bitwiseAnd(n int, k int) int {
    if (k - 1 | k) <= n {
        return k - 1
    }
    return k - 2
}

File: hackerrank/30 days of code/day 29, bitwise and/bitwise_and.rs
// https://www.hackerrank.com/challenges/30-bitwise-and/problem?isFullScreen=true

use text_io::{read, scan};

fn main() {
    let test_cases: usize = read!();
    for _ in 0..test_cases {
        let (n, k): (i32, i32);
        scan!("{} {}", n, k);
        println!("{}", bitwise_and(n, k));
    }
}

fn bitwise_and(n: i32, k: i32) -> i32 {
    if (k - 1 | k) <= n {
        k - 1
    } else {
        k - 2
    }
}
